broadlink/device.py

```diff
@@ -231,56 +278,198 @@ def set_lock(self, state: bool) -> None:
+
+                if nom_checksum != real_checksum:
+                    err = e.DataValidationError(
+                        -4008,
+                        "Received data packet check error",
+                        f"Expected a checksum of {nom_checksum} and received {real_checksum}",
+                    )
+                    _LOGGER.debug(err)
+                    errors.append(err)
+                    continue
+
+                err_code = int.from_bytes(resp[0x22:0x24], "little", signed=True)
+                resp_type = int.from_bytes(resp[0x26:0x28], "little")
+
+                if resp_type != exp_resp_type:
+                    err = e.DataValidationError(
+                        -4009,
+                        "Received data packet information type error",
+                        f"Expected {exp_resp_type} and received {resp_type}",
+                    )
+                    _LOGGER.debug(err)
+                    errors.append(err)
+                    continue
+
+                if not any(resp[:0x08]):
+                    return resp[0x30:], err_code
+
+                if resp[:0x08] != conn_id:
+                    continue
+
+                if len(resp) < 0x38:
+                    err = e.DataValidationError(
+                        -4010,
+                        "Received encrypted data packet length error",
+                        f"Expected at least 56 bytes and received {len(resp)}",
+                    )
+                    _LOGGER.debug(err)
+                    errors.append(err)
+                    continue
+
+                if int.from_bytes(resp[0x28:0x2A], "little") != count:
+                    continue
+
+                payload = resp[0x38:]
 
-        checksum = int.from_bytes(resp[0x20:0x22], "little")
-        if sum(resp, 0xBEAF) - sum(resp[0x20:0x22]) & 0xFFFF != checksum:
-            raise exception(-4008)  # Checksum error.
+                if len(payload) % 16:
+                    err = e.DataValidationError(
+                        -4010,
+                        "Received encrypted data packet length error",
+                        f"Expected a multiple of 16 and received {len(payload)}",
+                    )
+                    _LOGGER.debug(err)
+                    errors.append(err)
+                    continue
+
+                dev_ctrl_id = int.from_bytes(resp[0x30:0x34], "little")
+
+                if self.id and self.id != dev_ctrl_id:
+                    raise e.AuthorizationError(
+                        -4012,
+                        "Device control ID error",
+                        f"Expected {self.id} and received {dev_ctrl_id}",
+                    )
+
+                payload = self.decrypt(payload)
+                nom_p_checksum = int.from_bytes(resp[0x34:0x36], "little")
+                real_p_checksum = (
+                    p_checksum if err_code else sum(payload, 0xBEAF) & 0xFFFF
+                )
```

Comment:
There was a breaking change in the protocol at some point. The new firmware works like this:
```python3
real_p_checksum = sum(payload, 0xBEAF) & 0xFFFF
```
I need more time to investigate. I think it is reasonable to log and skip this error in v0.17.0.